File: t.py
# Author : ysh
# 2025/07/06 Sun 22:02:16
a, b = [int(i) for i in input().split()]
f = [int(i) for i in input().split()]

def check(x: int) -> bool:
    sig = 0
    for i in f:
        if i <= x:
            continue
        sig += (i - x)
    return sig >= b

def ck(l: int, r: int) -> int:
    if l + 1 == r:
        return l
    
    mid = (l + r) >> 1
    if check(mid):
        return ck(mid, r)
    else:
        return ck(l, mid)

print(ck(0, max(f)))

File: t.kt
// Author : ysh
// 2024/06/02 Sun 20:19:22
fun main() {
    var (a,b) = readLine()!!.split(' ').map {x -> x.toInt()};
    println("$a $b");
}
/*
fun main() {
    var (n,a,b) = readLine()!!.split(' ').map {it.toInt()}
    var f = (0 .. n - 1).toMutableList()

    fun ff(x: Int): Int = if(f[x] == x) x else {f[x] = ff(f[x]); f[x]}
    fun mg(a: Int, b: Int) {f[ff(a)] = ff(b); return}

    repeat(a) {
        var (a,b) = readLine()!!.split(' ').map {x -> x.toInt()}
        mg(a - 1,b - 1)
    }

    repeat(b) {
        var (a,b) = readln().split(' ').map {it.toInt()}
        println(if(ff(a - 1) == ff(b - 1)) "True" else "False")
    }
    return
}
*/
